e2e/api_test.go

```diff
@@ -0,0 +1,207 @@
+			"status/tasks/" + fakeSuccessTaskName,
+			map[string]api.TaskStatus{
+				fakeSuccessTaskName: api.TaskStatus{
+					TaskName:  fakeSuccessTaskName,
+					Status:    api.StatusHealthy,
+					Providers: []string{"fake-sync"},
+					Services:  []string{"api"},
+					EventsURL: "/v1/status/tasks/fake_handler_success_task?include=events",
+				},
+			},
+		},
+		{
+			"non-existing single task status",
+			"status/tasks/" + "non-existing-task",
+			map[string]api.TaskStatus{
+				"non-existing-task": api.TaskStatus{
+					TaskName:  "non-existing-task",
+					Status:    api.StatusUndetermined,
+					Providers: []string{},
+					Services:  []string{},
+					EventsURL: "",
+				},
+			},
+		},
+	}
+
+	for _, tc := range taskCases {
+		t.Run(tc.name, func(t *testing.T) {
+			u := fmt.Sprintf("http://localhost:%d/%s/%s", port, "v1", tc.path)
+			resp, err := http.Get(u)
+			require.NoError(t, err)
+			defer resp.Body.Close()
+
+			require.Equal(t, resp.StatusCode, http.StatusOK)
+
+			var taskStatuses map[string]api.TaskStatus
+			decoder := json.NewDecoder(resp.Body)
+			err = decoder.Decode(&taskStatuses)
+			require.NoError(t, err)
+
+			assert.Equal(t, tc.expected, taskStatuses)
+		})
+	}
+
+	overallCases := []struct {
+		name     string
+		path     string
+		expected api.OverallStatus
+	}{
+		{
+			"overall status",
+			"status",
+			api.OverallStatus{
+				Status: api.StatusCritical,
+			},
+		},
+	}
+
+	for _, tc := range overallCases {
+		t.Run(tc.name, func(t *testing.T) {
+			u := fmt.Sprintf("http://localhost:%d/%s/%s", port, "v1", tc.path)
+			resp, err := http.Get(u)
+			require.NoError(t, err)
+			defer resp.Body.Close()
+
+			require.Equal(t, resp.StatusCode, http.StatusOK)
+
+			var overallStatus api.OverallStatus
+			decoder := json.NewDecoder(resp.Body)
+			err = decoder.Decode(&overallStatus)
+			require.NoError(t, err)
+
+			assert.Equal(t, tc.expected, overallStatus)
+		})
+	}
+
+	err = stopCommand(cmd)
+	require.NoError(t, err)
+	removeDir(tempDir)
+}
+
+// runSyncDevMode runs the daemon in development which does not run or download
+// Terraform.
+func runSyncDevMode(configPath string) (*exec.Cmd, error) {
+	cmd := exec.Command("consul-terraform-sync",
+		fmt.Sprintf("--config-file=%s", configPath), "--client-type=development")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	return cmd, nil
+}
+
+func stopCommand(cmd *exec.Cmd) error {
+	cmd.Process.Signal(os.Interrupt)
+	sigintErr := errors.New("signal: interrupt")
+	if err := cmd.Wait(); err != nil && err != sigintErr {
+		return err
+	}
+	return nil
+}
+
+// registerService is a helper function to regsiter a service to the Consul
+// Catalog. The Consul sdk/testutil package currently does not support a method
+// to register multiple service instances, distinguished by their IDs.
+func registerService(t *testing.T, srv *testutil.TestServer, s testutil.TestService, health string) {
```

Comment:
This function is now duplicated in two packages, `e2e` and `templates/tftmpl`. It would be better to only use one instance of it in case it gets changed or removed to use the Consul testutil when supported, perhaps the `clients` package?

Reply:
Agree that it's duplicated in both packages and maintenance would have to happen in two places. I had tried reusing the method but didn't like how it looked and came from a place that prefers copying until something is used 3x so left it be... Moving it to a separate package sounds it would look better - let me try it.

Reply:
Moving to the client package looked nice but it unfortunately caused an import cycle. I forgot I had tried keeping the function in only e2e or templates/tftmpl but it led to an undefined error since both tests have separate build tags. Another option is to have the function in its own package. For me, it seems a little bit much for one function so I lean towards duplicating the function. Let me know if you have thoughts!

Reply:
That's unfortunate, thanks for looking into it! Yeah we can keep it as-is for now, until we have more test stragglers to group up.

Reply:
Of course! Agree - when we have some more cross-package test functions, we'll re-strategize. Thanks for the feedback!